### PR TITLE
SCP-32 SCP-33 SCP-34 SCP-40 User edit and auth on CRUD

### PR DESCRIPTION
**Pull request by jching1073**
User can now edit account (NOT PASSWORD)
AuthGuard applied to all relevant pages
Added login restrictions account, make survey and edit user in nav bar

### CODE REVIEW

File: server/controllers/user.ts [hunk 1]
```diff
@@ -59,6 +58,14 @@ export function displayRegisterPage(req:Request, res: Response, next: NextFuncti
 
 }
 
+export function displayUserEditPage(req:Request, res: Response, next: NextFunction): void {
+    if(!req.user)
```
Comment (JameeKim) [inline server/controllers/user.ts]: This part is not needed.

1. We should already have the authGuard middleware in our routes file. It should be guaranteed that the user is logged in at this point.
1. We also need to add a check on whether the requesting user is the same one with the requested id.
1. This route handler is not being used yet. Check the `server/routes/user.ts` file.
1. `!req.user` is the same as `!req.isAuthenticated()`, which returns true when the user is not logged in. This will result in not returning a response to the user request if the user is logged in.

File: server/controllers/user.ts [hunk 2]
```diff
@@ -124,3 +131,24 @@ export function processLogoutPage(req:Request, res: Response, next: NextFunction
 
     res.redirect("/login");
 }
+
+export function processEditPage(req: Request, res: Response, next: NextFunction): void {
```
Comment (JameeKim) [inline server/controllers/user.ts:139]: We need to check if the requesting user has the same id.

File: server/routes/survey.ts [hunk 3]
```diff
@@ -19,12 +19,13 @@ export default router;
...
 
 //new GET for surveyavailable:
 router.get("/surveyavailable", surveyController.displayAvailableSurvey);
 
 //NEW GET makesurvey page through displayMakeSurveyPage method
-router.get("/makesurvey", surveyController.displayMakeSurveyPage);
+router.get("/makesurvey", util.authguard ,surveyController.displayMakeSurveyPage);
 
 //POST - process makesurvey
 router.post("/makesurvey", surveyController.processMakeSurveyPage);
```
Comment (JameeKim) [inline server/routes/survey.ts]: We need the authGuard here, too.

File: server/routes/survey.ts [hunk 4]
```diff
@@ -36,16 +37,16 @@ router.get("/question/:id", surveyController.displayQuestionPage);
 router.post("/question/:id", responseController.processQuestion);
 
 //NEW GET question page through displayQuestionPage method
-router.get("/editsurvey/:id", surveyController.displayEditSurveyPage);
+router.get("/editsurvey/:id", util.authguard ,surveyController.displayEditSurveyPage);
 
 //POST- Process /edit/:id page
 router.post("/editsurvey/:id", surveyController.processEditSurveyPage);
```
Comment (JameeKim) [inline server/routes/survey.ts]: The authGuard should be here, also.

File: server/routes/user.ts [hunk 5]
```diff
@@ -33,9 +34,10 @@ router.post("/register", userController.processRegisterPage);
 router.get("/logout", userController.processLogoutPage);
 
 //NEW GET account page through displaydisplayAccountPage method
-router.get("/account", userController.displayAccountPage);
+router.get("/account", util.authguard ,userController.displayAccountPage);
 
 /* Get - display ediuser page - with /edituser */
-router.get("/edituser", (req, res, _next) => {
+router.get("/edituser/:id", (req, res, _next) => {
```
Comment (JameeKim) [inline server/routes/user.ts]: 1. We need to add the authGuard here.
2. Use the route handler created in the `server/controllers/user.ts` file.

File: server/controllers/user.ts [hunk 6]
```diff
@@ -124,3 +131,24 @@ export function processLogoutPage(req:Request, res: Response, next: NextFunction
...
+        "username": req.body.username,
+        "emailAddress": req.body.email,
+        "contactNumber": req.body.contactnumber
+    });
+    User.updateOne({_id: id}, updatedUser, {}, (err) =>{
+
+        if(err)
+        {
+            console.error(err);
+            res.end(err);
```
Comment (JameeKim) [inline server/controllers/user.ts]: There are 2 problems with this.

1. The callback function is not finished, so the `res.redirect` is still called in case there is an error. You would want to use `return` to end the callback function early to prevent that.
2. This sends a short description about the error to the user. By that, you're handling the error all by yourself at the place. But, we already have an error handler in place. You'd want to pass the error to that. To do so, you want to do `next(err)`.

In conclusion, what I suggest is like this:
```ts
if (err) {
    return next(err);
}
```

**Review by JameeKim — CHANGES_REQUESTED** [review]
Thanks for the work! There are a lot of comments coming from me because authentication is a very complicated thing to implement with a lot of edge cases.

One thing to note. There are currently 2 places that we need to check if the user sending the request is the same one as the user whose information is about to be edited: `displayUserEditPage` and `processEditPage`.

Other things, you can read through the comments below. :)

File: server/controllers/user.ts [hunk 7]
```diff
@@ -59,10 +59,14 @@ export function displayRegisterPage(req:Request, res: Response, next: NextFuncti
 }
 
 export function displayUserEditPage(req:Request, res: Response, next: NextFunction): void {
-    if(!req.user)
+
+    const id = req.params.id;
+
+    if(req.user?._id == id )
```
Comment (JameeKim) [inline server/controllers/user.ts]: I would suggest using `===` instead of `==`. It's just a javascript thing. `===` is the one that acts like `==` in most of the other languages.

File: server/controllers/user.ts [hunk 8]
```diff
@@ -59,10 +59,14 @@ export function displayRegisterPage(req:Request, res: Response, next: NextFuncti
...
     {
-        res.render("index", {title: "EditUser", page: "edituser"});
+        return res.render("index", {title: "EditUser", page: "edituser"});
     }
+    return res.redirect("/logout");
```
Comment (JameeKim) [inline server/controllers/user.ts]: I don't think we would want to log out the user. Maybe just redirect to the account page?

File: server/controllers/user.ts [hunk 9]
```diff
@@ -134,21 +138,24 @@ export function processLogoutPage(req:Request, res: Response, next: NextFunction
 
 export function processEditPage(req: Request, res: Response, next: NextFunction): void {
     const id = req.params.id;
+    if(req.user?._id == id)
```
Comment (JameeKim) [inline server/controllers/user.ts]: Same as above, use `===`.

Also, the `else` part is missing here. If we don't do anything, our server will not send anything to the user request. Maybe return `401 Unauthorized` or just redirect to the account page.

File: server/routes/user.ts [hunk 10]
```diff
@@ -33,9 +34,9 @@ router.post("/register", userController.processRegisterPage);
...
-router.get("/account", userController.displayAccountPage);
+router.get("/account", util.authguard ,userController.displayAccountPage);
 
 /* Get - display ediuser page - with /edituser */
-router.get("/edituser", (req, res, _next) => {
-    res.render("index", { title: "EditUser", page: "edituser" });
-});
+router.get("/edituser/:id", util.authguard ,userController.displayUserEditPage);
+
+router.post("/edituser/:id", userController.processEditPage);
```
Comment (JameeKim) [inline server/routes/user.ts]: AuthGuard?

**Review by JameeKim — CHANGES_REQUESTED** [review]
Thanks. Please check the new comments below and also the old ones that I didn't mark as resolved above.

**Review by JameeKim — APPROVED** [review]
Thank you so much!
